### PR TITLE
#182 Added SegmentPosition

### PR DESCRIPTION
**Pull request by nolmoonen**
#### Deployment instructions
None.

#### What have I done
Replaced `Route.segmentStartPositions` and `Route.segments` with `Route.segmentPositions`. `SegmentPosition` is a class comparable in functionality with `POIPosition`, in that is stores a reference to a segment and an offset on the route. Issue #182 .

#### Testing
Run unit tests.

### CODE REVIEW

File: src/app/backend/data-representation/route.ts [hunk 1]
```diff
@@ -29,99 +26,62 @@ export class Route {
...
+        if (this.segmentPositions === undefined || this.segmentPositions.length === 0) {
+            throw new IllegalStateError('segment positions not set');
         }
 
-        if (this.segmentStartPositions === undefined || this.segmentStartPositions.length === 0) {
-            throw new IllegalStateError('segmentStartPositions not set');
-        }
-
-        if (position < 0) {
+        if (position < 0 || position < this.getSegmentPositions()[0].getRouteOffset()) {
```
Comment (marijnvanderhorst) [inline src/app/backend/data-representation/route.ts]: Bit of inconsistent use of `segmentPositions`. Sometimes used by variable directly, sometimes via getter. I would say always use the variable here.

File: src/app/backend/file-parser/track-parser.service.spec.ts [hunk 2]
```diff
@@ -142,11 +143,11 @@ describe('TrackParserService', () => {
 
             path.setPath(segments, pois, 0);
             trackParserService.parse(trackText);
-            const orderedSegments: Segment[] = route.getSegments();
+            const orderedSegments = route.getSegmentPositions();
```
Comment (marijnvanderhorst) [inline src/app/backend/file-parser/track-parser.service.spec.ts]: Oh no, the type is gone.

File: src/app/backend/file-parser/track-parser.service.spec.ts [hunk 3]
```diff
@@ -200,10 +201,10 @@ describe('TrackParserService', () => {
 
         path.setPath(segments, pois, 0);
         trackParserService.parse(trackText);
-        const orderedSegments: Segment[] = route.getSegments();
+        const orderedSegments = route.getSegmentPositions();
```
Comment (marijnvanderhorst) [inline src/app/backend/file-parser/track-parser.service.spec.ts]: My precious types

File: src/app/backend/file-parser/track-parser.service.spec.ts [hunk 4]
```diff
@@ -255,10 +256,10 @@ describe('TrackParserService', () => {
 
             path.setPath(segments, pois, 0);
             trackParserService.parse(trackText);
-            const orderedSegments: Segment[] = route.getSegments();
+            const orderedSegments = route.getSegmentPositions();
```
Comment (marijnvanderhorst) [inline src/app/backend/file-parser/track-parser.service.spec.ts]: Not again

File: src/app/backend/file-parser/track-parser.service.spec.ts [hunk 5]
```diff
@@ -369,15 +369,15 @@ describe('TrackParserService', () => {
...
```
Comment (marijnvanderhorst) [inline src/app/backend/file-parser/track-parser.service.spec.ts]: I'm starting to see a trend here

File: src/app/backend/file-parser/track-parser.service.ts [hunk 6]
```diff
@@ -99,12 +99,18 @@ export class TrackParserService {
                     // poi was not in previous segment, add it so duplicates can be detected later
                     poisInPrevSegment.push(poiPosition.getPOI());
                 } else {
+                    const prevSegment = prevSegmentPosition.getSegment();
```
Comment (marijnvanderhorst) [inline src/app/backend/file-parser/track-parser.service.ts]: Please define this outside the for-loop.

File: src/app/backend/file-parser/track-parser.service.ts [hunk 7]
```diff
@@ -148,8 +158,21 @@ export class TrackParserService {
...
+                        let routeOffset: number;
+
+                        if (prevSegmentPosition) {
+                            // if previous segment position exists: previous offset + previous length
+                            routeOffset =
+                                prevSegmentPosition.getRouteOffset() +
+                                prevSegmentPosition.getSegment().getLength();
+                        } else {
+                            // if no previous offset exists, this segment is the first in route
+                            routeOffset = 0;
```
Comment (marijnvanderhorst) [inline src/app/backend/file-parser/track-parser.service.ts]: Why not let `routeOffset` be 0 initially and remove this entire else statement?

File: src/app/backend/data-representation/route.ts (hunk 1 above)
Comment (nolmoonen) [reply src/app/backend/data-representation/route.ts]: Changed.

File: src/app/backend/file-parser/track-parser.service.ts (hunk 6 above)
Comment (nolmoonen) [reply src/app/backend/file-parser/track-parser.service.ts]: Removed variable.

File: src/app/backend/file-parser/track-parser.service.ts (hunk 7 above)
Comment (nolmoonen) [reply src/app/backend/file-parser/track-parser.service.ts]: Good suggestion!

File: src/app/backend/file-parser/track-parser.service.spec.ts (hunk 2 above)
Comment (nolmoonen) [reply src/app/backend/file-parser/track-parser.service.spec.ts]: Added.

File: src/app/backend/file-parser/track-parser.service.spec.ts (hunk 3 above)
Comment (nolmoonen) [reply src/app/backend/file-parser/track-parser.service.spec.ts]: Added.

File: src/app/backend/file-parser/track-parser.service.spec.ts (hunk 4 above)
Comment (nolmoonen) [reply src/app/backend/file-parser/track-parser.service.spec.ts]: Added.

File: src/app/backend/file-parser/track-parser.service.spec.ts (hunk 5 above)
Comment (nolmoonen) [reply src/app/backend/file-parser/track-parser.service.spec.ts]: Added.

**Pull request merged**